gamemode/data/gamemoded.8.in
.\" Manpage for gamemoded.
.\" Contact projecttick@projecttick.org to correct errors or typos.
.TH gamemoded 8 "1 February 2026" "@GAMEMODE_VERSION@" "gamemoded man page"
.SH NAME
gamemoded \- daemon that optimises system performance on demand
.SH SYNOPSIS
\fBgamemoded\fR [OPTIONS...]
.SH DESCRIPTION
\fBGameMode\fR is a daemon/lib combo for Linux that allows games to request a set of optimisations be temporarily applied to the host OS.

The design has a clear cut abstraction between the host daemon and library (\fBgamemoded\fR and \fBlibgamemode\fR), and the client loaders (\fBlibgamemodeauto\fR and \fBgamemode_client.h\fR) that allows for safe usage without worrying about whether the daemon is installed or running. This design also means that while the host library currently relies on systemd for exchanging messages with the daemon, it's entirely possible to implement other internals that still work with the same clients.

\fBGameMode\fR was designed primarily as a stop-gap solution to problems with the Intel and AMD CPU powersave or ondemand governors, but is intended to be expanded beyond just CPU governor states, as there are a wealth of automation tasks one might want to apply.
.SH OPTIONS

.TP 8
.B \-r[PID], \-\-request=[PID]
Toggle gamemode for process.
When no PID given, requests gamemode and pauses
.TP 8
.B \-s[PID], \-\-status=[PID]
Query the status of gamemode for process
When no PID given, queries the status globally
.TP 8
.B \-d, \-\-daemonize
Run the daemon as a separate process (daemonize it)
.TP 8
.B \-l, \-\-log-to-syslog
Log to syslog
.TP 8 
.B \-h, \-\-help
Print help text
.TP 8
.B \-t, \-\-test
Run diagnostic tests on the current installation
.TP 8
.B \-v, \-\-version
Print the version

.SH USAGE
\fBlibgamemodeauto.so.0\fR can be pre-loaded into any program to request \fBgamemoded\fR begin or end the mode. See gamemoderun(1) for details.

The \fBgamemode_client.h\fR header can be used by developers to build the requests into a program:

.RS 4
.nf
#include "gamemode_client.h"

    if( gamemode_request_start() < 0 ) {
        fprintf( stderr, "gamemode request failed: %s\\n", gamemode_error_string() )
    }

    /* run the process */

    /* Not required, gamemoded can clean up after game exits */
    gamemode_request_end();
.fi
.RE

Atlernatively developers can define \fBGAMEMODE_AUTO\fR to mimic the behaviour of \fBlibgamemodeauto.so.0\fR:

.RS 4
.nf
#define GAMEMODE_AUTO
#include "gamemode_client.h"
.fi
.RE

Or, distribute \fBlibgamemodeauto.so.0\fR and either link with \fB\-lgamemodeauto\fR or inject it as above with \fBLD\_PRELOAD\fR.

.SH CONFIG

\fBgamemoded\fR will load and merge \fBgamemode.ini\fR config files from these directories in the following order:

.RS 4
@SYSCONFDIR@/
.RE
.RS 4
/etc/
.RE
.RS 4
$XDG_CONFIG_HOME or $HOME/.config/
.RE
.RS 4
$PWD
.RE

Behaviour of the config file can be explained by presenting a commented example:

.RS 4
.nf
@GAMEMODE_EXAMPLE_CONFIG@
.fi
.RE

.SH SEE ALSO
gamemoderun(1), systemd(1)

.SH ABOUT
GameMode source can be found at \fIhttps://github.com/FeralInteractive/gamemode.git\fR

.SH AUTHOR
Feral Interactive (linux-contact@feralinteractive.com)


gamemode/daemon/gamemoded.c
/*

Copyright (c) 2017-2025, Feral Interactive and the GameMode contributors
Copyright (c) 2026 Project Tick
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
 * Neither the name of Feral Interactive nor the names of its contributors
   may be used to endorse or promote products derived from this software
   without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

 */

/**
 * Simple daemon to allow user space programs to control the CPU governors
 *
 * The main process is responsible for bootstrapping the D-BUS daemon, caching
 * the initial governor settings, and then responding to requests over D-BUS.
 *
 * Clients register their pid(s) with the service, which are routinely checked
 * to see if they've expired. Once we reach our first actively registered client
 * we put the system into "game mode", i.e. move the CPU governor into a performance
 * mode.
 *
 * Upon exit, or when all clients have stopped running, we put the system back
 * into the default governor policy, which is invariably powersave or similar
 * on laptops. This ensures that the system is obtaining the maximum performance
 * whilst gaming, and allowed to sanely return to idle once the workload is
 * complete.
 */

#define _GNU_SOURCE

#include "gamemode.h"
#include "common-logging.h"
#include "gamemode-config.h"

#include "gamemode_client.h"

#include "build-config.h"

#include <fcntl.h>
#include <getopt.h>
#include <signal.h>
#include <sys/stat.h>
#include <systemd/sd-daemon.h> /* TODO: Move usage to gamemode-dbus.c */
#include <unistd.h>

#define USAGE_TEXT                                                                                 \
	"Usage: %s [-d] [-l] [-r] [-t] [-h] [-v]\n\n"                                                  \
	"  -r[PID], --request=[PID] Toggle gamemode for process\n"                                     \
	"                           When no PID given, requests gamemode and pauses\n"                 \
	"  -s[PID], --status=[PID]  Query the status of gamemode for process\n"                        \
	"                           When no PID given, queries the status globally\n"                  \
	"  -R, --reset              If gamemode is currently running, stop it, then restart\n"         \
	"  -d, --daemonize          Daemonize self after launch\n"                                     \
	"  -l, --log-to-syslog      Log to syslog\n"                                                   \
	"  -t, --test               Run tests\n"                                                       \
	"  -h, --help               Print this help\n"                                                 \
	"  -v, --version            Print version\n"                                                   \
	"\n"                                                                                           \
	"See man page for more information.\n"

#define VERSION_TEXT "gamemode version: v" GAMEMODE_VERSION "\n"

static void sigint_handler(__attribute__((unused)) int signo)
{
	LOG_MSG("Quitting by request...\n");
	sd_notify(0, "STATUS=GameMode is quitting by request...\n");

	/* Clean up nicely */
	game_mode_context_destroy(game_mode_context_instance());

	_Exit(EXIT_SUCCESS);
}

static void sigint_handler_noexit(__attribute__((unused)) int signo)
{
	LOG_MSG("Quitting by request...\n");
}

/**
 * Helper to perform standard UNIX daemonization
 */
static void daemonize(const char *name)
{
	/* Initial fork */
	pid_t pid = fork();
	if (pid < 0) {
		FATAL_ERRORNO("Failed to fork");
	}

	if (pid != 0) {
		LOG_MSG("Daemon launched as %s...\n", name);
		exit(EXIT_SUCCESS);
	}

	/* Fork a second time */
	pid = fork();
	if (pid < 0) {
		FATAL_ERRORNO("Failed to fork");
	} else if (pid > 0) {
		exit(EXIT_SUCCESS);
	}

	/* Now continue execution */
	umask(0022);
	if (setsid() < 0) {
		FATAL_ERRORNO("Failed to create process group\n");
	}
	if (chdir("/") < 0) {
		FATAL_ERRORNO("Failed to change to root directory\n");
	}

	/* replace standard file descriptors by /dev/null */
	int devnull_r = open("/dev/null", O_RDONLY);
	int devnull_w = open("/dev/null", O_WRONLY);

	if (devnull_r == -1 || devnull_w == -1) {
		LOG_ERROR("Failed to redirect standard input and output to /dev/null\n");
	} else {
		dup2(devnull_r, STDIN_FILENO);
		dup2(devnull_w, STDOUT_FILENO);
		dup2(devnull_w, STDERR_FILENO);
		close(devnull_r);
		close(devnull_w);
	}
}

/**
 * Main bootstrap entry into gamemoded
 */
int main(int argc, char *argv[])
{
	GameModeContext *context = NULL;

	/* Gather command line options */
	bool daemon = false;
	bool use_syslog = false;
	int opt = 0;

	/* Options struct for getopt_long */
	static struct option long_options[] = {
		{ "daemonize", no_argument, 0, 'd' },
		{ "log-to-syslog", no_argument, 0, 'l' },
		{ "request", optional_argument, 0, 'r' },
		{ "test", no_argument, 0, 't' },
		{ "status", optional_argument, 0, 's' },
		{ "reset", no_argument, 0, 'R' },
		{ "help", no_argument, 0, 'h' },
		{ "version", no_argument, 0, 'v' },
		{ NULL, 0, NULL, 0 },
	};
	static const char *short_options = "dls::r::tvhR";

	while ((opt = getopt_long(argc, argv, short_options, long_options, 0)) != -1) {
		switch (opt) {
		case 'd':
			daemon = true;
			break;
		case 'l':
			use_syslog = true;
			break;

		case 's':
			if (optarg != NULL) {
				pid_t pid = atoi(optarg);
				switch (gamemode_query_status_for(pid)) {
				case 0: /* inactive */
					LOG_MSG("gamemode is inactive\n");
					break;
				case 1: /* active not not registered */
					LOG_MSG("gamemode is active but [%d] not registered\n", pid);
					break;
				case 2: /* active for client */
					LOG_MSG("gamemode is active and [%d] registered\n", pid);
					break;
				case -1:
					LOG_ERROR("gamemode_query_status_for(%d) failed: %s\n",
					          pid,
					          gamemode_error_string());
					exit(EXIT_FAILURE);
				default:
					LOG_ERROR("gamemode_query_status returned unexpected value 2\n");
					exit(EXIT_FAILURE);
				}
			} else {
				int ret = 0;
				switch ((ret = gamemode_query_status())) {
				case 0: /* inactive */
					LOG_MSG("gamemode is inactive\n");
					break;
				case 1: /* active */
					LOG_MSG("gamemode is active\n");
					break;
				case -1: /* error */
					LOG_ERROR("gamemode status request failed: %s\n", gamemode_error_string());
					exit(EXIT_FAILURE);
				default: /* unexpected value eg. 2 */
					LOG_ERROR("gamemode_query_status returned unexpected value %d\n", ret);
					exit(EXIT_FAILURE);
				}
			}

			exit(EXIT_SUCCESS);

		case 'r':

			if (optarg != NULL) {
				pid_t pid = atoi(optarg);

				/* toggle gamemode for the process */
				switch (gamemode_query_status_for(pid)) {
				case 0: /* inactive */
				case 1: /* active not not registered */
					LOG_MSG("gamemode not active for client, requesting start for %d...\n", pid);
					if (gamemode_request_start_for(pid) < 0) {
						LOG_ERROR("gamemode_request_start_for(%d) failed: %s\n",
						          pid,
						          gamemode_error_string());
						exit(EXIT_FAILURE);
					}
					LOG_MSG("request succeeded\n");
					break;
				case 2: /* active for client */
					LOG_MSG("gamemode active for client, requesting end for %d...\n", pid);
					if (gamemode_request_end_for(pid) < 0) {
						LOG_ERROR("gamemode_request_end_for(%d) failed: %s\n",
						          pid,
						          gamemode_error_string());
						exit(EXIT_FAILURE);
					}
					LOG_MSG("request succeeded\n");
					break;
				case -1: /* error */
					LOG_ERROR("gamemode_query_status_for(%d) failed: %s\n",
					          pid,
					          gamemode_error_string());
					exit(EXIT_FAILURE);
				}

			} else {
				/* Request gamemode for this process */
				if (gamemode_request_start() < 0) {
					LOG_ERROR("gamemode request failed: %s\n", gamemode_error_string());
					exit(EXIT_FAILURE);
				}

				/* Request and report on the status */
				switch (gamemode_query_status()) {
				case 2: /* active for this client */
					LOG_MSG("gamemode request succeeded and is active\n");
					break;
				case 1: /* active */
					LOG_ERROR("gamemode request succeeded and is active but registration failed\n");
					exit(EXIT_FAILURE);
				case 0: /* inactive */
					LOG_ERROR("gamemode request succeeded but is not active\n");
					exit(EXIT_FAILURE);
				case -1: /* error */
					LOG_ERROR("gamemode_query_status failed: %s\n", gamemode_error_string());
					exit(EXIT_FAILURE);
				}

				/* Simply pause and wait a SIGINT */
				if (signal(SIGINT, sigint_handler_noexit) == SIG_ERR) {
					FATAL_ERRORNO("Could not catch SIGINT");
				}
				pause();

				/* Explicitly clean up */
				if (gamemode_request_end() < 0) {
					LOG_ERROR("gamemode request failed: %s\n", gamemode_error_string());
					exit(EXIT_FAILURE);
				}
			}

			exit(EXIT_SUCCESS);

		case 'R':
			switch (gamemode_request_restart()) {
			case 0: /* success */
				LOG_MSG("gamemode restart succeeded\n");
				exit(EXIT_SUCCESS);
			case 1: /* already off */
				LOG_ERROR("gamemode was already deactivated\n");
				break;
			case -1: /* error */
				LOG_ERROR("gamemode_request_restart failed: %s\n", gamemode_error_string());
				break;
			}
			exit(EXIT_FAILURE);

		case 't': {
			int status = game_mode_run_client_tests();
			exit(status);
		}
		case 'v':
			LOG_MSG(VERSION_TEXT);
			exit(EXIT_SUCCESS);
		case 'h':
			LOG_MSG(USAGE_TEXT, argv[0]);
			exit(EXIT_SUCCESS);
		default:
			fprintf(stderr, USAGE_TEXT, argv[0]);
			exit(EXIT_FAILURE);
		}
	}

	/* If syslog is requested, set it up with our process name */
	if (use_syslog) {
		set_use_syslog(argv[0]);
	}

	/* Daemonize ourselves first if asked */
	if (daemon) {
		daemonize(argv[0]);
	}

	/* Log a version message on startup */
	LOG_MSG("v%s\n", GAMEMODE_VERSION);

	/* Set up the game mode context */
	context = game_mode_context_instance();
	game_mode_context_init(context);

	/* Handle quits cleanly */
	if (signal(SIGINT, sigint_handler) == SIG_ERR) {
		FATAL_ERRORNO("Could not catch SIGINT");
	}
	if (signal(SIGTERM, sigint_handler) == SIG_ERR) {
		FATAL_ERRORNO("Could not catch SIGTERM");
	}

	/* Run the main dbus message loop */
	game_mode_context_loop(context);

	game_mode_context_destroy(context);

	/* Log we're finished */
	LOG_MSG("Quitting naturally...\n");
	sd_notify(0, "STATUS=GameMode is quitting naturally...\n");
}
